### ISSUE TITLE
documentation for how to build things

### ISSUE THREAD

**Issue by aeos** [expressjs/expressjs.com#222]
Hi,

I noticed that there is a push for better documentation, so I thought I would add an item that may be considered. Would it be possible to include documentation that explains the interface between the express application and view engine?

**Comment by aredridel** [comment]
+1

This would be excellent.

**Comment by dougwilson** [comment]
I believe this is already done: http://expressjs.com/faq.html

**Comment by aeos** [comment]
Hi Doug,

I believe the part that I had to guess at was the options object. Would it be possible to document what values the express application set such as "view cache", as well as proper use of the settings attribute and how locals is under _locals.

**Comment by dougwilson** [comment]
The settings are all in this table, including view cache: http://expressjs.com/4x/api.html#app-settings

I'm not sure what you mean by _locals; anything starting with a underscore is private and will never be documented.

**Comment by dougwilson** [comment]
And app.locals itself is documented here: http://expressjs.com/4x/api.html#app.locals

**Comment by dougwilson** [comment]
We'd be happy to document stuff, but so far it seems everything you are asking for is already documented. Perhaps the layout of the documentation is making hard for you to notice this? Do you have a suggestion for how we could make it easier for you to locate what you are looking for in the docs?

**Comment by dougwilson** [comment]
Also, if I'm still missing what you're asking for, feel free to post some code to point to it (you may simply be using terms to describe something I use different terms for).

**Comment by aeos** [comment]
Thanks for the quick replies, your answers made me go back through my past experiments to see where I may have been mistaken. The truth is that when I experimented with my view engine (https://github.com/aeos/nivlac) it was my first step into node and express. What confused me was the options object passed into __express and how it relates to the app.set, app.locals and res.locals. My desire was to pass an object into my view engine that contained only the view data. At first I thought that the locals object was the appropriate location. But the only way to know how to use it is to dump the options object. Unfortunately the locals object is merged with the settings (app.set) which was not exactly what I was trying to provide to the view engine. Then I discovered the _locals key in the options object and thought that maybe that was how it was accessed. 

My assumption is wrong, but the best practice is not always clear. In this case should I define a locals key called "ViewData" and then pass that key into my view engine?

One possible improvement to the guide is to add steps and best practices for developing each component. How do we best write middleware, or a view engine. Are there landmines such as using res.locals.settings. Are there standard approaches one should follow if one wants to release it to the world.

I dont know if anyone else may find this useful, as it is just one mans ideas.

Thanks for you time!
